ReactNativeClient/lib/services/KeymapService.ts

```diff
@@ -75,6 +78,36 @@ const defaultKeymap = {
 	],
 };
 
+export interface KeymapErrorDetails {
+	invalidCommand?: boolean,
+	invalidAccelerator?: boolean,
+	missingCommand?: boolean
+	missingAccelerator?: boolean,
+	invalidKeymapItem?: KeymapItem,
+	duplicateAccelerators?: boolean,
+	duplicateKeymapItems?: [KeymapItem, KeymapItem]
+}
+
+export class KeymapError extends Error {
+	private details_: KeymapErrorDetails;
+
+	constructor(message: string, details: KeymapErrorDetails = {}) {
+		super(message);
+		this.details_ = {
+			missingAccelerator: false,
+			missingCommand: false,
+			invalidAccelerator: false,
+			invalidCommand: false,
+			duplicateAccelerators: false,
+			...details,
+		};
+	}
+
+	get details() {
+		return this.details_;
+	}
```

Comment:
Could you extend JoplinError and simply set the code to some string? I think the details object with potentially multiple error causes seem unnecessarily complex. The keymap shortcut should be validated when the user inputs it anyway so we shouldn't have so much complexity going on. Check what they type, throw back an error with a message and that's almost all we need.

Reply:
I see. In that case I'll add some error handling to the ShortcutRecorder component. I'll update it.

Reply:
Is it okay to keep just an array of invalid Keymap items, so that we can show the little status icon next to the shortcut?

Reply:
What could cause the keymap to have multiple invalid items? If an item is invalid it shouldn't be possible to save it anyway, so there could only be one invalid item (the currently active, but unsaved, one), or am I missing something?

Reply:
No, you're right. We should prevent the user from causing the error by validating the shortcut before the user sets it. 
I'm working on it at the moment. I'll push an update in the next 24 hours.